### PR TITLE
added timeout parameter to the functions dir, ping, present, read, write

### PR DESCRIPTION
**Pull request by agentmnm**
communication with owserver gets aborted when the given timeout is exceeded
due to inner workings of owserver the granularity is in 1s steps only
e.g. if 0.5s is given as a timeout, the abortion will happen after 1s has elapsed

usage examples:

owproxy.read('sensA/temperature',2) # reading a temperature with timeout of 2s
owproxy.dir(timeout=1)          # reading directory with timeout of 1s

### CODE REVIEW

File: src/pyownet/protocol.py [hunk 1]
```diff
@@ -509,19 +532,20 @@ def dir(self, path='/', slash=True, bus=False):
         else:
             return []
 
-    def read(self, path, size=MAX_PAYLOAD, offset=0):
```
Comment (miccoli) [inline src/pyownet/protocol.py]: I think that it is better not to break the current API: `timeout=0` should always be the last argument. (Major changes should be delayed to forthcoming version 1.0.0)

File: src/pyownet/protocol.py [hunk 2]
```diff
@@ -509,19 +532,20 @@ def dir(self, path='/', slash=True, bus=False):
...
         ret, data = self.sendmess(MSG_READ, str2bytez(path),
-                                  size=size, offset=offset, )
+                                  size=size, offset=offset, timeout=timeout)
         if ret < 0:
             raise OwnetError(-ret, self.errmess[-ret], path)
         return data
 
-    def write(self, path, data, offset=0):
+    def write(self, path, data, timeout=0, offset=0):
         """write data at path
```
Comment (miccoli) [inline src/pyownet/protocol.py]: `timeout` should be last argument.

**Comment by miccoli** [comment]
Thanks for the pull request: I will look at the changes. I'm convinced that having timeouts is a very good idea, but I'm unsure if it is necessary to have a timeout argument for each method, instead of a global or per proxy instance constant value. 

For sure I will not change the current API: `read(path, 20)` now means `read(path, size=20)` and there is really no compelling reason to introduce a not backward compatible change.

File: src/pyownet/protocol.py (hunk 1 above)
Comment (agentmnm) [inline src/pyownet/protocol.py]: agreed

File: src/pyownet/protocol.py (hunk 2 above)
Comment (agentmnm) [inline src/pyownet/protocol.py]: ok

**Comment by agentmnm** [comment]
As for the API: I agree, changing it, was not a good idea. I was not aware that the method read can get called with the size parameter. Backward compatibility is more important than using timeout without labelling it in a call.

But for individual timeout per call basis vs. global: I think it could be possible to have both. There could be a max communication time an interaction with owserver can take per default - like 10s. At the same time there could be a timeout - set by the user e.g. me - on a per call basis that overwrites the global max com time. 

In my case I read 25 sensors in a row. A per call base timeout of 1s would ensure that, in case there are delays with the communication to one or two sensors, that the communication to the rest of the sensor would take place in a timely manner.

Let me know if I should work over this again and what you think of the changes in req and if it needs changes too.

**Comment by miccoli** [comment]
I will pull the proposed changes to a new branch and start working from there. Please stay tuned.

**Comment by agentmnm** [comment]
Okey dokey, thanks for the heads up.

**Comment by miccoli** [comment]
Pulled (with modifications) in 526b5a7356dc127788d444aa3bd35d2ffa3b6115

**Comment by agentmnm** [comment]
Cool! Thanks for improving my code and merging the timeout functionality on a per call basis.

**Comment by miccoli** [comment]
@agentmnm you are welcome. Thanks for the pull requests: I think that the per call timeout is much better than a global one

**Pull request merged**